common/lib/client/paginatedresource.ts

```diff
@@ -0,0 +1,195 @@
+			this.handlePage(err, body, headers, unpacked, statusCode, callback);
+		});
+	}
+
+	post<T1, T2>(params: Record<string, T2>, body: unknown, callback: PaginatedResultCallback<T1>): void {
+		Resource.post(this.rest, this.path, body, this.headers, params, this.envelope, (err: ErrorInfo, resbody: unknown, headers: Record<string, string>, unpacked: boolean, statusCode: number) => {
+			if(callback) {
+				this.handlePage(err, resbody, headers, unpacked, statusCode, callback);
+			}
+		});
+	}
+
+	put<T1, T2>(params: Record<string, T2>, body: unknown, callback: PaginatedResultCallback<T1>): void {
+		Resource.put(this.rest, this.path, body, this.headers, params, this.envelope, (err: ErrorInfo, resbody: unknown, headers: Record<string, string>, unpacked: boolean, statusCode: number) => {
+			if(callback) {
+				this.handlePage(err, resbody, headers, unpacked, statusCode, callback);
+			}
+		});
+	}
+
+	patch<T1, T2>(params: Record<string, T2>, body: unknown, callback: PaginatedResultCallback<T1>): void {
+		Resource.patch(this.rest, this.path, body, this.headers, params, this.envelope, (err: ErrorInfo, resbody: unknown, headers: Record<string, string>, unpacked: boolean, statusCode: number) => {
+			if(callback) {
+				this.handlePage(err, resbody, headers, unpacked, statusCode, callback);
+			}
+		});
+	}
+
+	handlePage<T>(err: ErrorInfo, body: unknown, headers: Record<string, string>, unpacked: boolean, statusCode: number, callback: PaginatedResultCallback<T>): void {
+		if(err && returnErrOnly(err, body, this.useHttpPaginatedResponse)) {
+			Logger.logAction(Logger.LOG_ERROR, 'PaginatedResource.handlePage()', 'Unexpected error getting resource: err = ' + Utils.inspectError(err));
+			callback(err);
+			return;
+		}
+		let items, linkHeader, relParams;
+		try {
+			items = this.bodyHandler(body, headers, unpacked);
+		} catch(e) {
+			/* If we got an error, the failure to parse the body is almost certainly
+			 * due to that, so cb with that in preference to the parse error */
+			callback(err || e);
+			return;
+		}
+
+		if(headers && (linkHeader = (headers['Link'] || headers['link']))) {
+			relParams = parseRelLinks(linkHeader);
+		}
+
+		if(this.useHttpPaginatedResponse) {
+			callback(null, new HttpPaginatedResponse(this, items, headers, statusCode, relParams, err));
+		} else {
+			callback(null, new PaginatedResult(this, items, relParams));
+		}
+	}
+}
+
+export class PaginatedResult <T> {
+	resource: PaginatedResource;
+	items: T[];
+	first?: (results: PaginatedResultCallback<T>) => void;
+	next?: (results: PaginatedResultCallback<T>) => void;
+	current?: (results: PaginatedResultCallback<T>) => void;
+	hasNext?: () => boolean;
+	isLast?: () => boolean;
+
+	constructor(resource: PaginatedResource, items: T[], relParams?: Record<string, any>) {
+		this.resource = resource;
+		this.items = items;
+
+		if(relParams) {
+			if('first' in relParams) {
+				this.first = (cb: (result?: ErrorInfo | null) => void) => {
```

Comment:
Ah. Now I can get what 'cb' is. Can this be named `callback`?

Reply:
I can't disagree with this. 😄 Though, perhaps, worth acknowledging that this variable name was inherited from the original JS implementation.

Reply:
500ea75

Reply:
Resolving this conversation on @ikbalkaya's behalf.